lib/facter/docker.rb

```diff
@@ -1,6 +1,34 @@
 require 'facter'
 require 'json'
 
+Facter.add(:docker_systemroot) do
+  confine :osfamily => :windows
+  setcode do
+    ENV['SystemRoot']
+  end
+end
+
+Facter.add(:docker_program_files_path) do
+  confine :osfamily => :windows
+  setcode do
+    ENV['ProgramFiles']
+  end
+end
+
+Facter.add(:docker_program_data_path) do
+  confine :osfamily => :windows
+  setcode do
+    ENV['ProgramData']
+  end
+end
+
+Facter.add(:docker_user_temp_path) do
+  confine :osfamily => :windows
+  setcode do
+    ENV['TEMP']
```

Comment:
For the Puppet service this is probably OK. For end users with Unicode characters in their usernames running `puppet apply`, there's a chance that this corrupts the directory name due to Ruby bugs.

If you must expose the TEMP directory as a fact (which I would generally advise against if you can help it), you'll want to use the built-in Puppet helpers `Puppet::Util.get_env` - see https://github.com/puppetlabs/puppet/blob/24ead48f617cd3912491fe419ac7b67cda53a320/lib/puppet/util.rb#L40-L52

Reply:
Thanks for the tip. I'll update the code with `Puppet::Util.get_env`.
Is there a reason why we should not use expose the TEMP directory as a custom fact and is there a better alternative to getting to the temporary directory from within a manifest?